schema_utils.py

```diff
@@ -387,3 +411,89 @@ def is_valid_email(obj):
             bool. Whether the given object is a valid email.
         """
         return bool(re.search(r'^[\w\.\+\-]+\@[\w]+\.[a-z]{2,3}$', obj))
+
+    @staticmethod
+    def is_valid_expression(obj, algebraic=True):
+        """Checks if the given  obj (a string) represents a valid algebraic or
+        numeric expression.
+
+        Args:
+            obj: str. The given expression.
+            algebraic: bool. True if the given expression is algebraic
+            else numeric.
+
+        Returns:
+            bool. Whether the given object is a valid expression.
+        """
+        valid_expression = True
+        valid_characters = '({[]})+-/*^.'
+
+        # Expression should not contain invalid characters.
+        for character in obj:
+            if character not in valid_characters and not bool(
+                    re.match(r'[a-zA-Z0-9]', character)):
+                valid_expression = False
+                break
+
+        # Algebraic expressions should contain at least one latin letter and
```

Comment:
Wouldn't numeric expressions be a subset of algebraic expressions (with regards to validity)?

Reply:
According to the definition that we finalized in the proposal, a numeric expression should not contain anything other than numbers and an algebraic expression should contain at least one variable. That's the only difference between the two.

Reply:
I don't remember this part. Could you point me to the discussion?

I agree that we need separate interactions for numeric and algebraic expressions. But even if the correct answer is 7x, the student can type 7. It seems to me that, logically, numeric expressions are a subset of algebraic expressions; the differences between the interactions are not just the range of inputs allowed, but also the rule checks.

Or are you planning to prevent students from typing "7"?

Reply:
I mentioned it in the warnings table [here](https://docs.google.com/document/d/1MkxE7mDPWZcY76v8HIz6W0LAQc5ckD6ozHQLVeR-69A/edit#bookmark=id.7xyychnrax4n). 

No, the student can't type in '7'. That would be considered as incorrect since if the creator has chosen the AlgebraicExpresionInput interaction then they would obviously be expecting an algebraic expression as a correct answer.
Also, in terms of validity, we require at least one variable to be present for an algebraic expression to be valid, so we can't use the same validity for numeric expressions since they can't contain variables.
I don't think numeric expressions are a subset of algebraic expressions. Also, the rule-set is pretty much the same for both (the only difference being that numeric expression input interaction won't have the MatchesWithGeneralForm rule, everything else is same).

Reply:
The link you provided doesn't support this discussion. It seems to point to the opposite case, which is entering a variable in the numeric expression input (which I agree should be disallowed).

That said, I think I'm fine going with what you propose for now. Just to check: if we find in the future that there are use cases where we e.g. want to give feedback on a specific wrong student answer to an "algebraic expression" question, and that student answer happens to be purely numeric, what steps would need to be taken to expand algebraic expression to cover numeric expressions and would this require any migrations?

Reply:
In that case, we would only have to remove this check from here and from the frontend and the creator can then add purely numeric expressions in an algebraic expression interaction as well.
I don't think this would need migrations since this wouldn't affect the already present data -- all of that would still be valid as they would all be algebraic expressions.

But note that the rule checks for the algebraic expression interaction would not work for numeric expressions (this was one of the reasons we decided to create separate interactions). They wouldn't necessarily throw errors if we remove this check, but they wouldn't work as expected either. For instance, 3+4 would be considered exactly equal to 7 (which is not what we want for an exact match). But yes, if the aim is just to detect if the learner has entered a purely numeric expression, then removing this check would be enough.
Also, note that the learner will get a helpful warning if they enter a numeric expression in the algebraic interaction.

Reply:
OK, thanks for the explanation. Let's leave the requirements as they are for now, then.